### ISSUE TITLE
Should we go ahead and add experimental `with` support?

### ISSUE THREAD

**Issue by seagreen** [pkamenarsky/concur-replica#16]
As thought up by you here: https://github.com/ajnsit/concur/issues/17#issuecomment-515750304

Without something like this `concur` is really hard to use for situations where you need widgets that both report values and maintain local state (without switching entirely to TEA).

We could mark it as HEAVILY experimental, so that it's there for people who are trying out the library, but people know not to rely on it.

**Comment by seagreen** [comment]
No idea if this is a good idea, but I thought it was at least worth discussing.

**Comment by pkamenarsky** [comment]
Yes, I agree it's worth experimenting with -- I think `with*` (or alternatively something else that solves the same problem) is essential to `Concur` programming, otherwise one must by necessity resort to TEA.

I'm not sure `concur-replica` is the right place for this though. Ideally `with` would be part of the main `Concur` repo, but I'm not sure @ajnsit is convinced of its usefulness. Iirc he is/was experimenting with `Wire`s and `Signal`s in the Purescript repo, but both abstractions seem to fall short of solving the problem `with` is solving.

So yeah, I'm fine with including those in `concur-replica` until either they're merged upstream or we come up with a better abstraction.

**Comment by seagreen** [comment]
Sounds great. I may take a stab at this after I finish SVG support.

A few thoughts:

+ I'm curious if we can make an API with read-only, write-only, and read-write options for the "Delta" vars, somewhat like Ömer Ağacan does in http://hackage.haskell.org/package/privileged-concurrency. I think that might be an easy way to make for more expressive types without adding much conceptual overhead.

+ I think good docs will be key for this to be useful, explaining the motivation behind it.

+ I definitely agree `concur-replica` shouldn't be the long term home for these functions.

**Comment by seagreen** [comment]
And another thoughts:

Do you remember why the "Delta" variable contains a channel? I wonder if we could get away with just a TVar.

**Comment by pkamenarsky** [comment]
Iirc, it's because multiple listeners could listen to the same "delta" - i.e. a global state which is shared among many independent components.

**Comment by seagreen** [comment]
Interesting. Multiple listeners can also react to a single TVar using `retry` to listen for changes, so it's possible we might be able to get away with just that. I'll have to do some experiments.

**Comment by pkamenarsky** [comment]
Yeah, but with `takeTVar`/`retry` the first woken up thread would steal the value. I think there's no getting around using channels unfortunately.

**Comment by seagreen** [comment]
> I think there's no getting around using channels unfortunately.

If so that's fine, it doesn't seem like a big cost.

However..

> Yeah, but with `takeTVar`/`retry` the first woken up thread would steal the value. I think there's no getting around using channels unfortunately.

Could the listening threads use `readTVar` instead of `takeTVar`? Using something like:

```hs
onChange :: Eq a => a -> TVar a -> STM a
onChange current var = do
  new <- readTVar var
  if current == new
    then
      retry
    else
      pure new
```

This seems a little more ""declarative"" since they would only be allowed to look at the current value, instead of being guaranteed to see every change.

I suppose it depends on the meaning of `Delta`. Do we want `Delta Int` to signify a stream of ints or the current value of that int?

Let me know what you think, I haven't tried this out yet so might be way off base.

**Comment by pkamenarsky** [comment]
That might actually work, and seems to be a lot simpler, although we'd need an `Eq` constraint here. Maybe there's a way to use `StablePtr` instead at the cost of some false positives, but this should be fine.

Also, I think `onChange` shouldn't take a `current` value as an argument, but rather get it from the `TVar` first before going into the loop.

**Comment by seagreen** [comment]
> Also, I think `onChange` shouldn't take a `current` value as an argument, but rather get it from the `TVar` first before going into the loop.

Nice, good idea!

**Comment by arthurxavierx** [comment]
I just described an alternative solution to this problem in https://github.com/ajnsit/concur/issues/17#issuecomment-877629202, please let me know what you think of it.
